File: app/models/main_force.py
"""
主力选股相关数据模型

定义主力选股功能的请求/响应模型和 MongoDB 文档模型。
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, ConfigDict, field_serializer
from bson import ObjectId
from .user import PyObjectId
from app.utils.timezone import now_tz


# ============================================================
# 请求/响应模型
# ============================================================


class MainForceScreenParams(BaseModel):
    """主力选股筛选参数"""
    time_range: str = Field(default="3m", description="时间区间: 3m/6m/1y/custom")
    start_date: Optional[str] = Field(default=None, description="自定义开始日期 YYYY-MM-DD")
    end_date: Optional[str] = Field(default=None, description="自定义结束日期 YYYY-MM-DD")
    top_n: int = Field(default=5, ge=3, le=10, description="精选推荐数量")
    max_change_pct: float = Field(default=30.0, description="最大涨跌幅限制(%)")
    min_market_cap: float = Field(default=50.0, description="最小市值(亿)")
    max_market_cap: float = Field(default=5000.0, description="最大市值(亿)")


class MainForceCandidate(BaseModel):
    """候选股票"""
    code: str = Field(..., description="股票代码")
    name: str = Field(..., description="股票名称")
    industry: str = Field(..., description="所属行业")
    net_inflow: float = Field(..., description="主力资金净流入(元)")
    change_pct: float = Field(..., description="区间涨跌幅(%)")
    market_cap: float = Field(..., description="总市值(亿)")
    pe_ratio: Optional[float] = Field(default=None, description="市盈率")
    pb_ratio: Optional[float] = Field(default=None, description="市净率")


class MainForceScreenResult(BaseModel):
    """筛选结果"""
    total_fetched: int = Field(..., description="获取的原始股票数量")
    total_filtered: int = Field(..., description="筛选后的股票数量")
    candidates: List[MainForceCandidate] = Field(default_factory=list, description="候选股票列表")


class MainForceAnalyzeRequest(BaseModel):
    """主力选股分析请求"""
    params: MainForceScreenParams = Field(default_factory=MainForceScreenParams, description="筛选参数")
    analysis_mode: str = Field(default="overview", description="分析模式: overview/batch")
    batch_count: Optional[int] = Field(default=None, ge=10, le=50, description="批量分析数量(10/20/30/50)")
    research_depth: str = Field(default="标准", description="研究深度")
    model_name: Optional[str] = Field(default=None, description="指定 LLM 模型名称（为空时自动选择）")
    selected_analysts: List[str] = Field(
        default_factory=lambda: ["fund_flow", "industry", "fundamental"],
        description="选择的分析师"
    )


class RecommendedStock(BaseModel):
    """推荐股票"""
    code: str = Field(..., description="股票代码")
    name: str = Field(..., description="股票名称")
    reason: str = Field(default="", description="推荐理由")
    highlights: List[str] = Field(default_factory=list, description="投资亮点")
    risks: List[str] = Field(default_factory=list, description="风险提示")
    position: str = Field(default="", description="建议仓位")
    period: str = Field(default="", description="投资周期")


class BatchResultItem(BaseModel):
    """批量分析单项结果"""
    code: str = Field(..., description="股票代码")
    name: str = Field(default="", description="股票名称")
    status: str = Field(default="pending", description="分析状态")
    rating: Optional[str] = Field(default=None, description="投资评级")
    confidence: Optional[float] = Field(default=None, description="信心度")
    entry_range: Optional[str] = Field(default=None, description="进场区间")
    take_profit: Optional[str] = Field(default=None, description="止盈位")
    stop_loss: Optional[str] = Field(default=None, description="止损位")
    target_price: Optional[str] = Field(default=None, description="目标价")
    advice: Optional[str] = Field(default=None, description="投资建议")
    error_message: Optional[str] = Field(default=None, description="错误信息")


# ============================================================
# MongoDB 文档模型
# ============================================================


class MainForceBatchHistory(BaseModel):
    """主力选股批量分析历史 - 对应 main_force_batch_history 集合"""
    id: Optional[PyObjectId] = Field(default_factory=PyObjectId, alias="_id")
    analysis_date: datetime = Field(default_factory=now_tz, description="分析日期")
    params: Dict[str, Any] = Field(default_factory=dict, description="筛选参数")
    candidates_count: int = Field(default=0, description="候选股票数量")
    filtered_count: int = Field(default=0, description="筛选后数量")
    recommended_count: int = Field(default=0, description="推荐股票数量")
    overview_analysis: Dict[str, str] = Field(
        default_factory=dict,
        description="各分析师报告: fund_flow_analyst, industry_analyst, fundamental_analyst, comprehensive_researcher"
    )
    recommended_stocks: List[Dict[str, Any]] = Field(
        default_factory=list,
        description="推荐股票列表"
    )
    batch_results: Dict[str, Any] = Field(
        default_factory=dict,
        description="批量分析结果: total, success, failed, duration_seconds, results"
    )
    task_id: Optional[str] = Field(default=None, description="关联的任务 ID")
    created_at: datetime = Field(default_factory=now_tz, description="创建时间")

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True
    )

    @field_serializer('analysis_date', 'created_at')
    def serialize_datetime(self, dt: Optional[datetime], _info) -> Optional[str]:
        """序列化 datetime 为 ISO 8601 格式，保留时区信息"""
        if dt:
            return dt.isoformat()
        return None

File: frontend/src/api/mainForce.ts
/**
 * 主力选股 API 模块
 *
 * 封装主力选股相关的所有 API 调用，包括筛选、分析、批量历史查询和删除。
 * 使用 ApiClient 统一请求封装。
 */

import { ApiClient } from './request'

// ============================================================
// 请求参数类型
// ============================================================

/** 主力选股筛选参数 */
export interface MainForceScreenParams {
  /** 时间区间: 3m/6m/1y/custom */
  time_range: string
  /** 自定义开始日期 YYYY-MM-DD */
  start_date?: string | null
  /** 自定义结束日期 YYYY-MM-DD */
  end_date?: string | null
  /** 精选推荐数量 (3-10) */
  top_n: number
  /** 最大涨跌幅限制(%) */
  max_change_pct: number
  /** 最小市值(亿) */
  min_market_cap: number
  /** 最大市值(亿) */
  max_market_cap: number
}

/** 主力选股分析请求 */
export interface MainForceAnalyzeRequest {
  /** 筛选参数 */
  params: MainForceScreenParams
  /** 分析模式: overview/batch */
  analysis_mode: string
  /** 批量分析数量(10/20/30/50) */
  batch_count?: number | null
  /** 研究深度 */
  research_depth: string
  /** 选择的分析师 */
  selected_analysts: string[]
  /** 指定 LLM 模型名称（为空时自动选择） */
  model_name?: string | null
}

// ============================================================
// 响应数据类型
// ============================================================

/** 候选股票 */
export interface MainForceCandidate {
  /** 股票代码 */
  code: string
  /** 股票名称 */
  name: string
  /** 所属行业 */
  industry: string
  /** 主力资金净流入(元) */
  net_inflow: number
  /** 区间涨跌幅(%) */
  change_pct: number
  /** 总市值(亿) */
  market_cap: number
  /** 市盈率 */
  pe_ratio?: number | null
  /** 市净率 */
  pb_ratio?: number | null
}

/** 筛选结果 */
export interface MainForceScreenResult {
  /** 获取的原始股票数量 */
  total_fetched: number
  /** 筛选后的股票数量 */
  total_filtered: number
  /** 候选股票列表 */
  candidates: MainForceCandidate[]
}

/** 分析响应 */
export interface MainForceAnalyzeResult {
  /** 整体分析任务 ID */
  task_id: string | null
  /** 批量分析 batch ID */
  batch_id?: string | null
  /** 候选股票数量 */
  candidates_count: number
  /** 获取的原始股票数量 */
  total_fetched: number
  /** 筛选后的股票数量 */
  total_filtered: number
}

/** 推荐股票 */
export interface RecommendedStock {
  code: string
  name: string
  reason: string
  highlights: string[]
  risks: string[]
  position: string
  period: string
}

/** 批量分析单项结果 */
export interface BatchResultItem {
  code: string
  name: string
  status: string
  rating?: string | null
  confidence?: number | null
  entry_range?: string | null
  take_profit?: string | null
  stop_loss?: string | null
  target_price?: string | null
  advice?: string | null
  error_message?: string | null
}

/** 批量分析历史记录 */
export interface MainForceBatchHistory {
  _id: string
  analysis_date: string
  params: Record<string, any>
  candidates_count: number
  filtered_count: number
  recommended_count: number
  overview_analysis: Record<string, string>
  recommended_stocks: RecommendedStock[]
  batch_results: {
    total: number
    success: number
    failed: number
    duration_seconds: number
    results: BatchResultItem[]
  }
  task_id?: string | null
  created_at: string
}

/** 批量历史分页响应 */
export interface BatchHistoryResponse {
  total: number
  page: number
  page_size: number
  items: MainForceBatchHistory[]
}

// ============================================================
// API 方法
// ============================================================

export const mainForceApi = {
  /**
   * 筛选主力资金股票
   * @param params 筛选参数
   */
  screen: (params: MainForceScreenParams) =>
    ApiClient.post<MainForceScreenResult>('/api/main-force/screen', params),

  /**
   * 触发完整分析流程（筛选 + AI 分析）
   * @param request 分析请求参数
   */
  analyze: (request: MainForceAnalyzeRequest) =>
    ApiClient.post<MainForceAnalyzeResult>('/api/main-force/analyze', request),

  /**
   * 查询批量分析历史
   * @param page 页码
   * @param pageSize 每页数量
   */
  getBatchHistory: (page = 1, pageSize = 20) =>
    ApiClient.get<BatchHistoryResponse>('/api/main-force/batch-history', {
      page,
      page_size: pageSize,
    }),

  /**
   * 删除指定的批量分析历史记录
   * @param historyId 历史记录 ID
   */
  deleteBatchHistory: (historyId: string) =>
    ApiClient.delete(`/api/main-force/batch-history/${historyId}`),
}
